src/DSROracleBase.sol

```diff
@@ -0,0 +1,138 @@
+        return _data.dsr;
+    }
+
+    function getChi() external view returns (uint256) {
+        return _data.chi;
+    }
+
+    function getRho() external view returns (uint256) {
+        return _data.rho;
+    }
+
+    function getAPR() external view returns (uint256) {
+        unchecked {
+            return (_data.dsr - RAY) * 365 days;
+        }
+    }
+
+    function getConversionRate() external view returns (uint256) {
+        return getConversionRate(block.timestamp);
+    }
+
+    function getConversionRate(uint256 timestamp) public view returns (uint256) {
+        IDSROracle.PotData memory d = _data;
+        uint256 rho = d.rho;
+        if (timestamp == rho) return d.chi;
+        require(timestamp >= rho, "DSROracleBase/invalid-timestamp");
+        
+        return (timestamp > rho) ? _rpow(d.dsr, timestamp - rho) * uint256(d.chi) / RAY : d.chi;
+    }
+
+    function getConversionRateBinomialApprox() external view returns (uint256) {
+        return getConversionRateBinomialApprox(block.timestamp);
+    }
+
+    // Copied and slightly modified from https://github.com/aave/aave-v3-core/blob/42103522764546a4eeb856b741214fa5532be52a/contracts/protocol/libraries/math/MathUtils.sol#L50
+    function getConversionRateBinomialApprox(uint256 timestamp) public view returns (uint256) {
+        IDSROracle.PotData memory d = _data;
+        uint256 rho = d.rho;
+        if (timestamp == rho) return d.chi;
+        require(timestamp >= rho, "DSROracleBase/invalid-timestamp");
+        
+        uint256 exp;
+        uint256 rate;
+        unchecked {
+            exp = timestamp - rho;
+            rate = d.dsr - RAY;
+        }
+
+        uint256 expMinusOne;
+        uint256 expMinusTwo;
+        uint256 basePowerTwo;
+        uint256 basePowerThree;
+        unchecked {
+            expMinusOne = exp - 1;
+
+            expMinusTwo = exp > 2 ? exp - 2 : 0;
+
+            basePowerTwo = rate * rate / RAY;
+            basePowerThree = basePowerTwo * rate / RAY;
+        }
+
+        uint256 secondTerm = exp * expMinusOne * basePowerTwo;
+        unchecked {
+            secondTerm /= 2;
+        }
+        uint256 thirdTerm = exp * expMinusOne * expMinusTwo * basePowerThree;
+        unchecked {
+            thirdTerm /= 6;
+        }
+
+        return d.chi * (RAY + (rate * exp) + secondTerm + thirdTerm) / RAY;
+    }
+
+    function getConversionRateLinearApprox() external view returns (uint256) {
+        return getConversionRateLinearApprox(block.timestamp);
+    }
+
+    function getConversionRateLinearApprox(uint256 timestamp) public view returns (uint256) {
+        IDSROracle.PotData memory d = _data;
+        uint256 rho = d.rho;
+        if (timestamp == rho) return d.chi;
+        require(timestamp >= rho, "DSROracleBase/invalid-timestamp");
+        
+        uint256 duration;
+        uint256 rate;
+        unchecked {
+            duration =  timestamp - rho;
+            rate = uint256(d.dsr) - RAY;
```

Comment:
```suggestion
            duration = timestamp - rho;
```

Reply:
Why are you flipping these?

Reply:
This went through wrong, just meant to say to remove extra space after `duration =`

Reply:
fixed